aioxmpp/pep/service.py

```diff
@@ -0,0 +1,238 @@
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+########################################################################
+import asyncio
+import contextlib
+
+import aioxmpp
+import aioxmpp.service as service
+import aioxmpp.callbacks as callbacks
+
+
+class PEPClient(service.Service):
+    """
+    :class:`PEPClient` simplifies working with PEP services.
+
+    Compared to :class:`~aioxmpp.PubSubClient` it supports automatic
+    checking for server support, a stream-lined API. It is intended to
+    make PEP things easy. If you need more fine-grained control or do
+    things which are not usually handled by the defaults when using PEP, use
+    :class:`~aioxmpp.PubSubClient` directly.
+
+    See :class:`register_pep_node` for the high-level interface for
+    claiming a PEP node and receiving event notifications.
+
+    There also is a low-level interface:
+
+    .. automethod:: claim_pep_node
+
+    .. automethod:: unclaim_pep_node
+
+    """
+    ORDER_AFTER = [
+        aioxmpp.DiscoClient,
+        aioxmpp.DiscoServer,
+        aioxmpp.PubSubClient
+    ]
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._pubsub = self.dependencies[aioxmpp.PubSubClient]
+        self._disco_client = self.dependencies[aioxmpp.DiscoClient]
+        self._disco_server = self.dependencies[aioxmpp.DiscoServer]
+
+        self._pep_node_claims = {}
+
+    def claim_pep_node(self, node_namespace, handler, *,
+                       register_feature=True, notify=False):
+        """
+        Claim node `node_namespace`.
+
+        Dispatch event notifications for `node_namespace` to `handler`.
+
+        This registers `node_namespace` as feature for service discovery
+        unless ``register_feature=False`` is passed.
+
+        :param node_namespace: the pubsub node whose events shall be
+            handled.
+        :param handler: the handler to install.
+        :type handler: callable, see
+            :attr:`aioxmpp.PubSubClient.on_item_publish`
+            for the arguments.
+        :param register_feature: Whether to publish the `node_namespace`
+            as feature.
+        :param notify: Whether to register the ``+notify`` feature to
+            receive notification without explicit subscription.
+
+        :raises RuntimeError: if a handler for `node_namespace` is already
+            set.
+        """
```

Comment:
I prefer to have (in docstrings):
```python
"""
One paragraph (ideally one line) summary.

:param foo: …
:raises fnord: if baz
:return: bar

Detailed description
"""
```

The advantage is that the level of detail increases section by section and that the reader has a relatively constant offset from the synopsis to the overview of what the parameters do. This is a shortcoming of the official python docs imo (you always have to search to find the docs for a specific parameter).